radio-active: add module to create config files

S0AndS0 opened the pull request:
**WARN:**

- must be applied after https://github.com/NixOS/nixpkgs/pull/441029
- requires update of `flake.lock` too

### Description

Configuration entry similar to;

```nix
programs.radio-active.enable = true;
```

...  will produce default configuration `~/.radio-active-configs.ini`

```
[AppConfig]
filepath = /home/{user}/recordings/radioactive/
filetype = mp3
filter = none
limit = 100
loglevel = info
player = ffplay
sort = name
volume = 80
```

By default `ffplay` is used for recording/playback, but that can be changed by applying either of the following;

```nix
programs.radio-active.appConfig.player = "vlc";
programs.radio-active.appConfig.player = "mpv";
```

All other configuration options documented by;
https://github.com/deep5050/radio-active?tab=readme-ov-file#default-configs maybe applied under the `AppConfig` attribute set.

Finally, the `aliases` attribute set allows for defining key/value pares that will generate a `~/.radio-active-alias` of bookmarked stations, for example something like;

```nix
programs.radio-active.aliases = {
  "Deep House Lounge" = "http://198.15.94.34:8006/stream";
};
```

... will result in;

```
Deep House Lounge==http://198.15.94.34:8006/stream
```

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [ ] Change is backwards compatible.
   > Nope

- [x] Code formatted with `nix fmt` or
      `nix-shell -p treefmt nixfmt deadnix keep-sorted --run treefmt`.

- [x] Code tested through `nix run .#tests -- test-all` or
      `nix-shell --pure tests -A run.all`.
   > Note; had to do `--impure` and import local `pkgs` due to pending PR over at [NixOS/nixpkgs#441029](https://github.com/NixOS/nixpkgs/pull/441029), but _should_ work once that's accepted and `flake.lock` is updated here ;-)

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [x] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [x] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [x] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

rycee commented on modules/programs/radio-active.nix:
```diff
@@ -0,0 +1,188 @@
+          };
+        };
+
+    home.file."radio-active-alias" = mkIf (cfg.settings != { } && cfg.settings.aliases != { }) {
+      text = ''
+        ${builtins.concatStringsSep "\n" (
+          mapAttrsToList (name: value: "${name}==${value}") cfg.settings.aliases
+        )}
+      '';
+      target = "${config.home.homeDirectory}/.radio-active-alias";
```
No need for the `target` attribute:

```suggestion
    home.file.".radio-active-alias" = mkIf (cfg.settings != { } && cfg.settings.aliases != { }) {
      text = ''
        ${builtins.concatStringsSep "\n" (
          mapAttrsToList (name: value: "${name}==${value}") cfg.settings.aliases
        )}
      '';
```

rycee commented on modules/programs/radio-active.nix:
```diff
@@ -0,0 +1,188 @@
+
+      type = submodule {
+        options = {
+          AppConfig = mkOption {
```
I would suggest the following structure instead:

- Make `programs.radio-active.settings` match this `AppConfig` option.
- Move the `aliases` option to the top.

With the above a valid configuration could be

``` nix
{
  enable = true;
  settings.limit = 30;
  aliases = {
    "Deep House Lounge" = "http://198.15.94.34:8006/stream";
  };
}
```

S0AndS0 replied:
The configuration file could have more than `AppConfig` bits defined, configs like;

```ini
[AppConfig]
# ... stuff for "AppConfig" settings...
```

...  map to a Python dictionary sorta like `configs['AppConfig']['player']` -> `"ffplay"`

So lifting `AppConfig` into the `config.programs.radio-active.settings` attribute set could lead to future churn when (if) additional sections are added to the configuration file.

Currently is not an issue but, as I'm currently the fork maintainer submitted to `NixOS/nixpkgs` for the program, I also hesitate to reduce future flexibility.

rycee replied:
Fair enough, then the settings can include the `AppConfig` attribute. Can just move the aliases option.

rycee replied:
That's correct, basically you should be able to change

```nix
            source = iniFormat.generate "radio-active-config" {
              inherit (cfg.settings) AppConfig;
            };
```

to

```nix
            source = iniFormat.generate "radio-active-config" cfg.settings;
```

S0AndS0 replied:
Noice, that's gonna simplify future maintenance _some_ 

:warning: I'm rebasing onto `master` to get tests passing locally, because it now seems possible to simplify the `settings.type -> submodule {/*...*/}` stuff, so there's gonna be _some_ `git diff` churn due to jumping onto commit `11cc3d55ded3346a8195000ddeadde782a611e56`

S0AndS0 commented:
Hmm; recent force-push, after satisfying linter, seems to show that Darwin don't got `pkgs.vlc`?

> `https://buildbot.nix-community.org/api/v2/logs/1487024/raw_inline`

> ...
> `error: Package ‘vlc-3.0.21’ in /nix/store/yw0wijix50n0y2ii15sk8kvnwgalpq6f-source/pkgs/by-name/vl/vlc/package.nix:322 is not available on the requested hostPlatform:`
> ...

Grep-ing about via `grep -rinE 'platform' tests/modules` shows others, like `tests/modules/services/osmscout-server/default.nix`, are leveraging `pkgs.stdenv.hostPlatform.isLinux` to make certain tests conditional.  So that's what I'm gonna do for `tests/modules/programs/radio-active/default.nix` until told otherwise.

rycee commented on modules/programs/radio-active.nix:
```diff
@@ -0,0 +1,194 @@
+              }
+            )
+          else
+            lib.warn ''
+              Unknown player defined in `config.programs.radio-active.AppConfig.player`
+            '' pkgs.radio-active;
```
Warnings should go in the `warnings` option, that way they are presented consistently. See, e.g., https://github.com/nix-community/home-manager/blob/173a29f735c69950cfeaac310d7e567115976be0/modules/programs/bun.nix#L51-L55

S0AndS0 commented:
> Thanks! I added a commit to simplify the aliases and settings examples, squashed and merged to master now 🙂

Woot, thanks for the guidance and extra help!
